### DISCUSSION TITLE
Remove Fizzler Dependency

### DISCUSSION THREAD

**Discussion by inforithmics** [vvvv/SVG#1073] [Ideas]
I think I could make a Pull request to remove the Fizzler dependency, because this is the only place where it is used.

`var elemsToStyle = rootNode.QuerySelectorAll(rule.Selector.Text, elementFactory);`

And here the css selector is again parsed with fizzler, the rule.Selector is already a parsed css selector.

So this would have following benefits:
1. Faster (css Selector is only parsed once).
2. The Fizzler dependency could be removed.

**Comment by inforithmics** [comment]
I had a look at it, but it seems that CombinatorSelector needs to be made public. in Excss

**Reply by inforithmics** [reply]
I "discovered" the Licensing issue

https://github.com/svg-net/SVG/issues/563

Alternatives to Fizzler AngleSharp But I think for that IElement needs to be implemented on SvgElement.

https://github.com/wieslawsoltes/Svg.Skia/issues/162

Use excss CssSelector for evaluation.
https://github.com/TylerBrinks/ExCSS/pull/160

I made a POC (Proof of concept) and in my first 10 tests 50% pass. Maybe AngleSharp is the better approach but it is a very Big dependency.

**Comment by mrbean-bremen** [comment]
The licensing issue is not a problem IMO (I have closed the issue again, because the open issue of using ExCSS as a library has long been fixed). The author of the library has no problem with the usage, and the license permits the usage as is - so I see no real reason to replace it.
Removing the dependeny at all would be another matter though...

**Reply by inforithmics** [reply]
Here it says that LGPL Libraries need to be replaceable by users. But with current Developments in the .NET space that .NET Programs are linked together. This would violate the LGPL. In Mobile Applications like Xamarin, Maui, ... where the Applikations are Linked for saving space and techically a User cannot change libraries, this would probably be a violation too.

https://en.wikipedia.org/wiki/GPL_linking_exception

My POC with replacing Fizzler passes all Image Comparision Tetsts, so Basic CSS works. But I discovered a regression in ExCSS 4.2.1 ExCss 4.2.0 works.

So I will make a Pull Request for fixing that first.

**Reply by mrbean-bremen** [reply]
From my understanding, the [LGPL](https://en.wikipedia.org/wiki/LGPL) allows to link to LGPL code from code with another license, as long as the LGPL code is not modified (which it isn't in our case). I'm not a lawyer and may be wrong, but I have seen LGPL libraries used in several propritary software where the legal department was quite strict to make sure that no license is violated (and have been involved in license clearing myself) - so I doubt that it would really be a problem.

Anyway, I'm not against making SVG dependent on less libraries, I'm just not convinced that there is a licensing problem. There certainly was one before we removed the copied Fizzler source code from SVG, though.

**Reply by inforithmics** [reply]
I'm no Legal Person. 
But my main motiviation is to reduce the dependencies. And ExCss parses the Selectors and Fizzler parses the Same Selectors which is in my opinion redundant and only ExCss is needed for the Css Handling.

Pull Request for updating to newest ExCss Version:
https://github.com/svg-net/SVG/pull/1083

Pull Request which was merged for Exposing Dom from the Selector for evaluation:
https://github.com/TylerBrinks/ExCSS/pull/160

The Pull Request for removeing Fizzler and Evaluation can only be made when ExCss releases a new Version.

**Reply by inforithmics** [reply]
A link to LGPL and Apple Ecosystem
https://softwareengineering.stackexchange.com/questions/291229/xamarin-mono-lgpl-compliance

**Reply by inforithmics** [reply]
Here a exhaustive link. 

Main take aways lgpl should be dynamic linked (I think the .Net Mobile Applications are static linked especially since they have code removed and static compiled (AOT)).
If they are static linked the developer that uses a lgpl library has to provide tools for static linking a different version of the library.

https://fossa.com/blog/open-source-software-licenses-101-lgpl-license/

**Reply by mrbean-bremen** [reply]
Ah ok, thanks. I had missed the part about .Net Mobile Applications being statically linked, that would indeed be a problem. I don't have any experience in that area, so I overlooked it...
